docs/dev/metrics.md

```diff
@@ -0,0 +1,53 @@
+# CAO Metrics
+
+The Cluster Autoscaler Operator reports the following metrics:
+
+## Metrics provided by the controller runtime
+
+The [controller runtime](https://github.com/kubernetes-sigs/controller-runtime)
+integration with the operator provides metrics about the webhook admission server.
+You can find more information about these metrics names and their labels through
+the following links:
+
+### Kubernetes controller metrics
+
+The labels `controller="cluster_autoscaler_controller"` and
+`controller="machine_autoscaler_controller"` can be used to refine queries against these metrics.
+* [Controller runtime reconciliation metrics implementation](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/internal/controller/metrics/metrics.go)
+
+### Admission webhook metrics
+
+The label `webhook="/validate-clusterautoscalers"` can be used to refine the
+queries for these metrics.
+* [Controller runtime webhook metrics implementation](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/webhook/internal/metrics/metrics.go)
+
+### Prometheus REST server metrics
+
+The `url` label can be quite useful for querying these metrics, here are a few
+of the available URL values to use: `"https://172.30.0.1:443/%7Bprefix%7D"`, `"https://172.30.0.1:443/apis?timeout=32s"`.
```

Comment:
From where should I utilize these URLs?  I'm assuming these IPs are subject to change?

Reply:
these are the in-pod IP addresses that are assigned to the prometheus http server, they are used as labels during promql calls to isolate the different endpoints. in my tests it looked like these IP addresses always came up the same, but i have to imagine it would change if the address range changed for the container network.

Reply:
Just as a heads up, I believe `172.30.0.1` is the service IP of the Kubernetes service in the default namespace on an OpenShift cluster, ie, it's the service that fronts the API server, not sure why this is coming up WRT prometheus though

Reply:
is it possible we are just seeing a local ip address that only exists within the pod network for these containers?

Reply:
The pod network on openshift is a `10.x` subnet, the `172.30.x` is used for services, perhaps the metrics are for requests that CAO is making to the Kubernetes API?

Reply:
could be, i'll take some time to dig in a little and see if i can learn more here.

i guess another way to look at this is, should we just remove this info and point to the docs in the prometheus project? my concern is that we get focused on this metric when it seems like a minor thing.

Reply:
This is a valid concern, I agree, maybe we just drop that little bit

Reply:
first, thank you both for pushing a little on this one it's been a fascinating investigation. as near as i can tell, these metrics are being generated by client-go using the controller-runtime package. i'm pretty sure [this is where it starts](https://github.com/kubernetes/client-go/blob/master/rest/request.go#L720). it appears to be showing all the outbound requests that are being made by the client-go inside the operator. which would square up with what Joel is saying about the IP address being the root API server.

so, maybe explain that bit here and give the example IP address? or is that too much detail?